File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
// bfs로 가능한(=이어진) 경로 모두 탐색
// 레퍼런스 참조 - https://velog.io/@sean2337/Programmers-%EC%A0%84%EB%A0%A5%EB%A7%9D%EC%9D%84-%EB%91%98%EB%A1%9C-%EB%82%98%EB%88%84%EA%B8%B0-JavaScript
function solution(n, wires) {
  let answer = Infinity
  
  // 0. 트리 뼈대 만들고, 채우기 - n+1 -> 1번인덱스에는 1번이 들어가도록 하기 위해
  let tree = Array.from({length: n+1}, () => [])
  for (let wire of wires) {
    let [from, to] = wire
    tree[from].push(to);
    tree[to].push(from);
  }

  // 1. bfs함수 만들기 - 인자: start, 제외할 번호
  function bfs(start, exceptNum) {
    let cnt = 0;
    let visit = []; // n만큼 0 안채워도 됨 ㅇㅇ.
    let queue = [start];
    visit[start] = true;

    while (queue.length) {
      let idx = queue.pop()
      // 반복문 말고, forEach써보자 - 안 익숙해서 일부러 안썼는데 연습ㄱㄱ
      tree[idx].forEach((ele) => {
        if (ele !== exceptNum && visit[ele] !== true) {
          visit[ele] = true;
          queue.push(ele);
        }
      })
      cnt ++
    }

    return cnt
  }

  wires.forEach(ele => {
    let[a,b] = ele
    // 아래 부분이 제일 이해x였음 - 왜 bfs(a,b)-bfs(b,a)인가?
    // 문제 홈페이지에서 입출력예시1 그림 보면 바로 이해가 됨
    // bfs(start,except)에서 start는 root노드임 -> 여기서 쭉 뻗어나가는 거고,
    // except는 root에서 해당 번호쪽으로는 안 가겠다는 것
    // 따라서, bfs(a,b) - bfs(b,a)는 a와b의 연결이 끊어진 상태에서 각각의 전력망 개수의 차이를 구한것과 같다
    answer = Math.min(answer, Math.abs(bfs(a,b)-bfs(b,a))) 
  })

  return answer
}

console.log(solution(9,[[1,3],[2,3],[3,4],[4,5],[4,6],[4,7],[7,8],[7,9]]))

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
# 그런데, 3분할 되는 경우는 없는가? -> 문제에선 2분할만 예정한 듯 보임
    # "당신은 이 전선들 중 하나를 끊어서 현재의 전력망 네트워크를 2개로 분할..."
# 흠;; 그냥.. dfs인가..? -> 정확히는 bfs
#  wires목록중에 요소 1개 없애도록 순회 돌고 - 맨 밖에서
    # 안에서 bfs함수 돌리는 것
    # bfs는 check만 하도록
# wire들의 연결 정보는 dict에 key-value에 저장하는게 나을듯
    # *근데 이 부분의 코드가 잘 안 짜짐 - key로 value값 찾아서 들어가지만 다시 나오는 경우가 특히.
    # https://wikidocs.net/196182보고 힌트깨달음 -> 일단 연결되었으면 다 value에 적어주면 되겠구나
        # 이에 따르면 이번 그래프는 '무방향'그래프
    # *[1,2]면 -> 1:2, 2:1이렇게 두개 하는거임 swap해서! - 있으면 추가x식으로~

    # 걍 dict하지말고({1:2} x) -> [1,2]해서 in으로 찾을까?
# visisited배열 선언해서, key, value둘다 접근한 경우 false->true로 바꿈
from collections import deque


def bfs(grp, length): # 인자로 뭐를 넣어야 할까
    connected = 0
    visited = [False] * (length+1) # 1인덱스가 1송전탑 방문여부
    queue = deque()
    now_key = list(grp.keys())[0]
    if not grp[now_key]:
        return length-1 # 여기서 끝
    visited[now_key] = True # key: 방문표시
    connected += 1

    queue.append(grp[now_key]) # [3,4,5]리스트 형태로 들어감 / 첫 시작

    while queue:
        now_values = queue.popleft()

        for value in now_values:
            if not visited[value]: # False상태면
                visited[value] = True
                connected += 1
                queue.append(grp[value])
        
    return abs(2*connected-length) # 2a-length가 a-b값임


def solution(n, wires):
    result = []

    for i in range(len(wires)):
        tmp = wires[:]
        tmp.pop(i)

        graph = {x: [] for x in range(1,n+1)} # key는 1부터 시작
        
        for [one,two] in tmp: # 이렇게도 되네 [one,two] zz
            if two not in graph[one]:
                graph[one].append(two)
            if one not in graph[two]:
                graph[two].append(one)
        result.append(bfs(graph,n))
    
    return min(result)

print(solution(9,[[1,3],[2,3],[3,4],[4,5],[4,6],[4,7],[7,8],[7,9]]))
print(solution(4,[[1,2],[2,3],[3,4]]))
print(solution(7,[[1,2],[2,7],[3,7],[3,4],[4,5],[6,7]]))